object/object.go

```diff
@@ -121,3 +131,31 @@ func SplitSlabs(slabs []Slab, lengths []int) [][]SlabSlice {
 func SingleSlabs(slabs []Slab, length int) []SlabSlice {
 	return SplitSlabs(slabs, []int{length})[0]
 }
+
+type rekeyStream struct {
+	key []byte
+	c   *chacha20.Cipher
+
+	counter uint64
+	nonce   uint64
+}
+
+func (rs *rekeyStream) XORKeyStream(dst, src []byte) {
+	rs.counter += uint64(len(src))
+	if rs.counter < 64*math.MaxUint32 {
+		rs.c.XORKeyStream(dst, src)
+		return
+	}
+	// counter overflow; xor remaining bytes, then increment nonce and xor again
+	rem := 64*math.MaxUint32 - (rs.counter - uint64(len(src)))
```

Comment:
```go
var rem uint64
if rs.counter < 64*math.MaxUint32 {
    rem = 64*math.MaxUint32-rs.counter
}

if uint64(len(src)) <= rem {
    rs.counter += uint64(len(src))
    rs.c.XORKeyStream(dst, src)
    return
}

rs.counter = uint64(len(src)) - rem
rs.c.XORKeyStream(dst[:rem], src[:rem])

// NOTE: we increment the last 8 bytes because XChaCha uses the
// first 16 bytes to derive a new key; leaving them alone means
// the key will be stable, which might be useful.
rs.nonce++
nonce := make([]byte, 24)
binary.LittleEndian.PutUint64(nonce[16:], rs.nonce)
rs.c, _ = chacha20.NewUnauthenticatedCipher(rs.key, nonce)
rs.c.XORKeyStream(dst[rem:], src[rem:])
```

`nit` Original also fine I just checked the math and started typing an example and found it was more readable (to me at least) written out like this.